chanacceptor/chainedacceptor.go

```diff
@@ -61,11 +67,135 @@ func (c *ChainedAcceptor) Accept(req *ChannelAcceptRequest) *ChannelAcceptRespon
+				req.OpenChanMsg.PendingChannelID, err)
+
+			return NewChannelAcceptResponse(
+				false, errChannelRejected, nil, 0, 0,
+				0, 0, 0, 0,
+			)
+		}
 	}
 
 	// If we have gone through all of our acceptors with no objections, we
 	// can return an acceptor with a nil error.
-	return NewChannelAcceptResponse(true, nil)
+	return &finalResp
+}
+
+// mergeResponse takes two channel accept responses, and attempts to merge their
+// fields, failing if any fields conflict (are non-zero and not equal). It
+// returns a new response that has all the merged fields in it.
+func mergeResponse(current, new ChannelAcceptResponse) (ChannelAcceptResponse,
+	error) {
+
+	val, err := mergeValue(
+		"csv delay", current.CSVDelay == 0, new.CSVDelay == 0,
+		current.CSVDelay, new.CSVDelay,
+	)
+	if err != nil {
+		return current, err
+	}
+	current.CSVDelay = val.(uint16)
+
+	val, err = mergeValue(
+		"htlc limit", current.HtlcLimit == 0, new.HtlcLimit == 0,
+		current.HtlcLimit, new.HtlcLimit,
+	)
+	if err != nil {
+		return current, err
+	}
+	current.HtlcLimit = val.(uint16)
+
+	val, err = mergeValue(
+		"min htlc in", current.MinHtlcIn == 0,
+		new.MinHtlcIn == 0, current.MinHtlcIn,
+		new.MinHtlcIn,
+	)
+	if err != nil {
+		return current, err
+	}
+	current.MinHtlcIn = val.(lnwire.MilliSatoshi)
+
+	val, err = mergeValue(
+		"min depth", current.MinAcceptDepth == 0,
+		new.MinAcceptDepth == 0, current.MinAcceptDepth,
+		new.MinAcceptDepth,
+	)
+	if err != nil {
+		return current, err
+	}
+	current.MinAcceptDepth = val.(uint16)
+
+	val, err = mergeValue(
+		"in flight total", current.InFlightTotal == 0,
+		new.InFlightTotal == 0, current.InFlightTotal,
+		new.InFlightTotal,
+	)
+	if err != nil {
+		return current, err
+	}
+	current.InFlightTotal = val.(lnwire.MilliSatoshi)
+
+	val, err = mergeValue(
+		"reserve", current.Reserve == 0, new.Reserve == 0,
+		current.Reserve, new.Reserve,
+	)
+	if err != nil {
+		return current, err
+	}
+	current.Reserve = val.(btcutil.Amount)
+
+	val, err = mergeValue(
+		"upfront shutdown", current.UpfrontShutdown == nil,
+		new.UpfrontShutdown == nil, current.UpfrontShutdown,
+		new.UpfrontShutdown,
+	)
+	if err != nil {
+		return current, err
+	}
+	current.UpfrontShutdown = val.(lnwire.DeliveryAddress)
+
+	return current, nil
+}
+
+// mergeValue checks that a current value for a field and the new proposed
+// value do not contradict each other, and returns the new value that should
+// be set.
+func mergeValue(name string, currentEmpty bool, newEmpty bool,
+	current, new interface{}) (interface{}, error) {
+
+	switch {
+	// If the current value is empty, then we can set any new value we want.
+	case currentEmpty:
+		return new, nil
+
+	// If there is no new value proposed, then we can just stick with the
+	// old value.
+	case newEmpty:
+		return current, nil
+
+	// If both values are not empty, but are not equal, then we have an
+	// inconsistent set of values.
+	case current != new:
```

Comment:
not positive, but my intuition says that direct interface comparisons could have some unexpected behavior. for example, what happens when merging two non-nil `UpfrontShutdownScripts`? in this playground example, attempting to compare two interfaced byte slices causes a runtime panic https://play.golang.org/p/cC53YdpsCwW. maybe i'm missing something that prevents this?

Reply:
TIL, great catch! Updated to just use separate merge functions for the different types, sharing where possible. Also added tests for the merge function so that everything is covered.